lib/modules/datasource/github-releases/cache-base.ts

```diff
@@ -0,0 +1,369 @@
+   * How many days to wait until full cache reset (for single package).
+   */
+  resetAfterDays?: number;
+
+  /**
+   * How many days ago the package should be published to be considered as stable.
+   * Since this period is expired, it won't be refreshed via soft updates anymore.
+   */
+  unstableDays?: number;
+
+  /**
+   * How many items per page to obtain per page during initial fetch (i.e. pre-fetch)
+   */
+  itemsPerPrefetchPage?: number;
+
+  /**
+   * How many pages to fetch (at most) during the initial fetch (i.e. pre-fetch)
+   */
+  maxPrefetchPages?: number;
+
+  /**
+   * How many items per page to obtain per page during the soft update
+   */
+  itemsPerUpdatePage?: number;
+
+  /**
+   * How many pages to fetch (at most) during the soft update
+   */
+  maxUpdatePages?: number;
+}
+
+/**
+ * The options that are meant to be used in production.
+ */
+const cacheDefaults: Required<CacheOptions> = {
+  /**
+   * How many minutes to wait until next cache update
+   */
+  updateAfterMinutes: 30,
+
+  /**
+   * How many days to wait until full cache reset (for single package).
+   */
+  resetAfterDays: 7,
+
+  /**
+   * How many days ago the package should be published to be considered as stable.
+   * Since this period is expired, it won't be refreshed via soft updates anymore.
+   */
+  unstableDays: 30,
+
+  /**
+   * How many items per page to obtain per page during initial fetch (i.e. pre-fetch)
+   */
+  itemsPerPrefetchPage: 100,
+
+  /**
+   * How many pages to fetch (at most) during the initial fetch (i.e. pre-fetch)
+   */
+  maxPrefetchPages: 100,
+
+  /**
+   * How many items per page to obtain per page during the soft update
+   */
+  itemsPerUpdatePage: 100,
+
+  /**
+   * How many pages to fetch (at most) during the soft update
+   */
+  maxUpdatePages: 100,
+};
+
+/**
+ * Tells whether the time `duration` is expired starting
+ * from the `date` (ISO date format) at the moment of `now`.
+ */
+function isExpired(
+  now: DateTime,
+  date: string,
+  duration: DurationLike
+): boolean {
+  const then = DateTime.fromISO(date);
+  const expiry = then.plus(duration);
+  return now >= expiry;
+}
+
+export abstract class AbstractGithubDatasourceCache<
+  StoredItem extends StoredItemBase,
+  FetchedItem = unknown
+> {
+  private updateDuration: DurationLike;
+  private resetDuration: DurationLike;
+  private stabilityDuration: DurationLike;
+
+  private maxPrefetchPages: number;
+  private itemsPerPrefetchPage: number;
+
+  private maxUpdatePages: number;
+  private itemsPerUpdatePage: number;
+
+  constructor(private http: GithubHttp, opts: CacheOptions = {}) {
+    const {
+      updateAfterMinutes,
+      resetAfterDays,
+      unstableDays,
+      maxPrefetchPages,
+      itemsPerPrefetchPage,
+      maxUpdatePages,
+      itemsPerUpdatePage,
+    } = {
+      ...cacheDefaults,
+      ...opts,
+    };
+
+    this.updateDuration = { minutes: updateAfterMinutes };
+    this.resetDuration = { days: resetAfterDays };
```

Comment:
Can we add some random minutes / hours, so not all caches expirre at same time? Otherwise we have massive github queries on same renovate run. especially for self-hostet users.